elements/pl-units-input/unit_utils.py

```diff
@@ -0,0 +1,152 @@
+
+import numpy as np
+import prairielearn as pl
+from pint import Quantity, UnitRegistry
+from typing_extensions import assert_never
+
+# TODO write tests for each of these functions
+
+CORRECT_UNITS_INCORRECT_MAGNITUDE_FEEDBACK = (
+    "Your answer has correct units, but incorrect magnitude."
+)
+INCORRECT_UNITS_CORRECT_MAGNITUDE_FEEDBACK = (
+    "Your answer has correct magnitude, but incorrect units."
+)
+INCORRECT_UNITS_AND_MAGNITUDE_FEEDBACK = (
+    "Your answer has incorrect units and magnitude."
+)
+INCORRECT_FEEDBACK = "Your answer is incorrect."
+
+
+class ComparisonType(Enum):
+    RELABS = "relabs"
+    SIGFIG = "sigfig"
+    EXACT = "exact"
+    DECDIG = "decdig"
+
+
+def get_only_units_grading_fn(
+    *, ureg: UnitRegistry, correct_ans: str
+) -> Callable[[str], Tuple[bool, Optional[str]]]:
+    """Returns the grading function used for units only grading mode."""
+    parsed_correct_ans = ureg.Quantity(correct_ans)
+
+    def grade_only_units(submitted_ans: str) -> Tuple[bool, Optional[str]]:
+        parsed_submission = ureg.Quantity(submitted_ans)
+        if parsed_correct_ans.units == parsed_submission.units:
+            return True, None
+
+        return False, INCORRECT_FEEDBACK
+
+    return grade_only_units
+
+
+def get_exact_units_grading_fn(
+    *,
+    ureg: UnitRegistry,
+    correct_ans: str,
+    comparison: ComparisonType,
+    magnitude_partial_credit: Optional[float],
+    digits: int,
+    rtol: float,
+    atol: str,
+) -> Callable[[str], Tuple[float, Optional[str]]]:
+    parsed_correct_ans = ureg.Quantity(correct_ans)
+    parsed_atol = ureg.Quantity(atol)
+
+    def magnitude_comparison_fn(
+        submitted_magnitude: float, correct_magnitude: float
+    ) -> bool:
+        submitted_magnitude_parsed = np.float64(submitted_magnitude)
+        correct_magnitude_parsed = np.float64(correct_magnitude)
+
+        """Returns true if submitted_magnitude is close enough to correct_magnitude based on comparison algorithm"""
+        if comparison is ComparisonType.EXACT:
+            return submitted_magnitude_parsed == correct_magnitude_parsed
+        elif comparison is ComparisonType.SIGFIG:
+            return pl.is_correct_scalar_sf(
+                a_sub=submitted_magnitude_parsed,
+                a_tru=correct_magnitude_parsed,
+                digits=digits,
+            )
+        elif comparison is ComparisonType.DECDIG:
+            return pl.is_correct_scalar_dd(
+                a_sub=submitted_magnitude_parsed,
+                a_tru=correct_magnitude_parsed,
+                digits=digits,
+            )
+        elif comparison is ComparisonType.RELABS:
+            return pl.is_correct_scalar_ra(
+                a_sub=submitted_magnitude_parsed,
+                a_tru=correct_magnitude_parsed,
+                rtol=rtol,
+                atol=parsed_atol.magnitude,
+            )
+
+        assert_never(comparison)
+
+    magnitudes_match_credit = (
```

Comment:
I would just use an `if` statement if the ternary is going to be split across multiple lines.

Reply:
I think this was a one-liner before, but the autoformatter broke it up.